sources/engine/Stride.Engine/Engine/Startup/GameBuilder.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) .NET Foundation and Contributors (https://dotnetfoundation.org/ & https://stride3d.net)
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Reflection;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+
+namespace Stride.Engine.Startup;
+
+public interface IGameBuilder
+{
+    public ConfigurationManager Configuration { get; }
+    public IServiceCollection Services { get; }
+    public ILoggingBuilder Logging { get; }
+    public IAppEnvironment Environment { get; }
+
+    IStrideApplication Build();
+}
+
+internal class GameBuilder : IGameBuilder
+{
+    public GameBuilder()
+    {
+        Configuration = new ConfigurationManager();
+
+        Environment = new AppEnvironment { ApplicationName = Assembly.GetEntryAssembly()?.GetName().Name ?? "", EnvironmentName = GetEnvironment(), ContentRootPath = AppContext.BaseDirectory };
+        
+        var physicalFileProvider = new PhysicalFileProvider(Environment.ContentRootPath);
+        Environment.ContentRootFileProvider = physicalFileProvider;
+        Configuration.SetFileProvider(physicalFileProvider);
+
+        // As an example of how ASP handles appsettings.Development.json
+        // overriding configuration values from appsettings.json
+        Configuration
+            .AddJsonFile("appsettings.json", optional: true)
+            .AddJsonFile($"appsettings.{Environment.EnvironmentName}.json", optional: true);
+
+        Services = new ServiceCollection();
+
+
+        Services.AddSingleton(Environment);
+        Services.AddSingleton(_ => Configuration);
+
+        Services.AddLogging();
+        Logging = new LoggingBuilder(Services);
+    }
+    
+    public ConfigurationManager Configuration { get; }
+    public IServiceCollection Services { get; }
+    public ILoggingBuilder Logging { get; }
+    public IAppEnvironment Environment { get; }
+    
+    private static string GetEnvironment()
+    {
+#if DEBUG
+        return Environments.Development;
+#elif STAGING
+        return Environments.Staging;
+#endif
+        return Environments.Production;
+    }
```

Comment:
Perhaps more control could be given back to the developers. The `ASPNETCORE_ENVIRONMENT` environment variable is used for those kinds of application. A custom environment variable of our own to specify the game environment could be nice. It would also give devs more flexibility for custom environment types they may have. I do believe that expecting `Development` or `Production` as reserved values would be a good idea since we would want to use them for extension methods like `.IsDevelopment()`.

Reply:
Yeah looking back at this now, not happy with this anymore, especially since I learned that the whole idea of using `Environment` is to specifically _not_ use the msbuild `$(Configuration)`.